New Structure & Contents (#31)
blindnet-io/blindnet.dev: docs/02-tutorial.mdx

---
sidebar_position: 2
title: Tutorial
tags: [quickstart]
---

:::info

This tutorial doesn’t assume any existing knowledge of the blindnet devkit or even any privacy-related technologies.

:::

The purpose of the following tutorial is to serve as an entry point for the blindnet devkit, by simply walking you through its different components to give you a better understanding of the overall structure and logic.

> 

Reviewer: If we indeed assume no prior knowlegde about the devkit, then before the next paragraph (where we tell them that we'll aswer questions on the purpose of Privacy Computation Engine) it seams appropriate to add here a brief overview of what the components are, or at least list them.

Reviewer: Maybe even an image showing all the components would be appropriate here.

Author: I am indeed writing a better overview explainer as we speak (and it does include a graph).

However, I didn't decide if it should be part of the main introduction or a separate document yet (it will depend on how long it has to be).

I consider the introductory tutorial should just give a "quick pick" overview, so I especially want to avoid going too deep into conceptual and technical details. It shouldn't take more than 10 minutes to be completed, or max 20 minutes if absolutely necessary

As soon as the devkit is ready enough, the introductory tutorial will only be a "learn by practice" guide (to create a first complete "privacy stack" step by step, using all components), while more advanced tutorials will allow going deeper into each component separately.

Doing so, we'll answer the following questions:

> - What is the purpose of the Privacy Computation Engine;

Reviewer: For all the linkable documentation concepts (like here Privacy Computation Engine), I would suggest to adopt the same rule as on wikipedia. The first time the concept is mentioned in a text (or a section of a text if it is expected that sections might be read independently), the concept is linked to the corresponding documentation page. 

Every next time, it is just noted in the First Letter Upper Case notation without beining linked.

3. **Development**: call the Privacy Computation Engine from your registered application to handle Privacy Requests

Step _(1) App Management_ and _(2) Configuration_ don't need to be addressed for this tutorial.

:::tip Pre-configured Instance

To make things easier for this tutorial, we have already pre-configured a dedicated custom instance for you to play with:

<div class="text--center">

**`https://devkit-pce-staging.azurewebsites.net/v0/`**

</div>

:::

Step _(3) Development_ can be achieved directly by calling the standalone Privacy Computation Engine [REST API](https://devkit-pce-staging.azurewebsites.net/swagger/).

However, the true strength of the blindnet devkit comes when combining its various components, thus drastically improving your developer experience and productivity while giving you access to additional features.

Starting with our premade user interfaces.

<!-- ## 2. Capture Data -->
<!-- TODO: add information on how to capture data without the Capture Componet, i.e. call the demo storage service API to save a data from a simple HTML form -->

## 2. Add Privacy-Rights Management Capabilities to Your Application

> User interfaces for the blindnet devkit (aka _[Privacy Components for the Web](https://github.com/blindnet-io/privacy-components-web)_) are delivered as standard Web Components available on both npm and a CDN.

Reviewer: Would it be possible to provide a link for Web Components, to a page (maybe wikipedia) where it is explained what they are?

Author: Addressed in https://github.com/blindnet-io/blindnet.dev/pull/40

As a consequence, installing the blindnet devkit web interfaces doesn't involve any specificity, as you can install and use them as you would with any other standard Web Component.

:::info

See [Using Web Components](/docs/interfaces/using-web-components) for more information on how to integrate Web Components in your existing project.

:::

In the interest of simplicity and clarity, let's see how to integrate these components into a simple static HTML page, starting with the Privacy Request Capture Interface (PRCI).

> PRCI is a simple interface allowing your users to assert their privacy rights by sending you a simple yet standardized [Privacy Request](/docs/references/lexicon#privacy-request). It is an entry point for your user to privacy, not only allowing you to cover GDPR and other privacy-related legislations but also and more importantly, to build trust by showing your users their privacy rights are tangibly handled, enforced and respected.

Reviewer: ```suggestion
PRCI is a simple interface allowing your users to assert their privacy rights by sending you a simple yet standardized [Privacy Request](/docs/references/lexicon#privacy-request). 
By implementing it you will improve your compliance with GDPR and other privacy-related legislations. 
Also, more importantly, studies show that complying with privacy regulations and facilitating Privacy Requests improves trust and helps build stronger relationships with users. 
```

Author: With developers in mind, I believe we should really avoid things like "studies show" and promises like "you will improve" without a direct answer to "why" questions. I'd prefer these phrasings to be limited to the "references" section, blindnet.io, and our more generic communication (blog, LinkedIn, etc.).

The sentence you suggested to change is way too long tho. I'll try to find a better phrasing.

Author: Alternative suggestion:

```suggestion
PRCI is an intuitive interface allowing your users to assert their privacy rights by sending you a simple yet standardized [Privacy Request](/docs/references/lexicon#privacy-request).
As such, it gives your users a direct entry point to their privacy.

Adding this component to your web application helps you improve your compliance with GDPR and other privacy-related legislations.
Also, more importantly, the PRCI will directly improve your users' trust in your organization by showing them their privacy rights are tangibly handled, enforced and respected.
```

> To integrate this interface into your web page, you'll first need to load the associated bundle (included in the `@blindnet/prci` npm package) from the unpkg CDN by adding a `<script>` tag to your HTML document:

Reviewer: ```suggestion
To integrate this interface into your web page, you can load the associated bundle (included in the `@blindnet/prci` npm package) from the unpkg CDN by adding a `<script>` tag to your HTML document:
```
Invitational language instead of a conditional

Author: I'd totally agree with you in a more generic context. But this is a tutorial, so it's still better to formulate it in a directive way ("do this, do that"), which means invitational language should be avoided.

```html
<script src="https://unpkg.com/@blindnet/prci/dist/index.core.min.js?module" type="module"></script>
```

:::info

Here, we use the `core` module build for performance improvement.

As this build only integrates the _core_ parts of the library, external dependencies like [Lit](https://lit.dev) will only be loaded once, even after we add other privacy components to this same web page.

Your specific context might be different. This is why each Privacy Component for the Web includes different builds and bundles covering any specific needs, as detailed in the [User Interfaces - Install](/docs/interfaces/install) section.

:::

> Then, you only need to add the associated entry-point custom element `<bln-priv-request>` to your HTML document:

Reviewer: ```suggestion
Then, you may want to add the associated entry-point custom element `<bln-priv-request>` to your HTML document:
```
Invitational language

Author: see https://github.com/blindnet-io/blindnet.dev/pull/31#discussion_r966945868

```html
<h2 id="priv-request">Make a privacy request</h2>

<bldn-priv-request></bldn-priv-request>
```

Leading us to the following HTML file:

```html
<!DOCTYPE html>
<html lang="en-GB">
  <head>
    <meta charset="utf-8" />
    <title>Privacy Components for the Web - Minimal Static Demo</title>
  </head>
  <body>
    <h1>Privacy Components for the Web - Minimal Static Demo</h1>

    <h2>Assert Your Privacy Rights</h2>

    <bldn-priv-request></bldn-priv-request>

    <script src="https://unpkg.com/@blindnet/prci/dist/index.core.min.js?module" type="module" ></script>
  </body>
</html>
```

:::info Demo

> This simple static demo can be tested live on [blindnet-io.github.io](https://blindnet-io.github.io/privacy-components-web/demos/devkit-simple-tutorial/demos/static/).

Reviewer 2: Page not found for this url

Author: That's indeed an error of inattention.

```suggestion
This simple static demo can be tested live on [blindnet-io.github.io](https://blindnet-io.github.io/privacy-components-web/demos/static/).
```

BTW, this demo is broken on `main` currently, but it has been fixed in the next release (see https://github.com/blindnet-io/privacy-components-web/pull/112).

> Code can be found in the [privacy-components-web repository](ttps://github.com/blindnet-io/privacy-components-web/tree/main/demos/static).

Reviewer 2: ```suggestion
Code can be found in the [privacy-components-web repository](https://github.com/blindnet-io/privacy-components-web/tree/main/demos/static).
```

You can either directly clone the whole monorepo locally:

```bash
git clone git@github.com:blindnet-io/privacy-components-web.git
cd privacy-components-web/demos/devkit-simple-tutorial/
```

or use [git sparse-checkout](https://git-scm.com/docs/git-sparse-checkout) to only download the project itself:

```bash
mkdir devkit-web-demo
cd devkit-web-demo
git init
git remote add -f origin git@github.com:blindnet-io/privacy-components-web.git
git config core.sparsecheckout true
echo "demos/devkit-simple-tutorial" >> .git/info/sparse-checkout
git pull origin main
```

Then, you'll need to install the project's dependencies using classic yarn (v1) or npm:

```bash
yarn install
# OR
npm install
```

Then, start the development server:

```bash
yarn start
# OR
npm start
```

and open the application in your browser using the given URL.

:::info

The next steps have been already developed in this demo project. You can therefore refer to it whenever in doubt.

:::

## 3. Assert Your Privacy Rights

Now, open the demo web app or static page in your browser to see this component in action:

![first screen of the PRCI component](/img/captures/prci-entry.png)

> Here, you can see your users can directly choose from a wide range of privacy request types.

Reviewer: ```suggestion
Here, your users can directly choose from a wide range of privacy request types.
```

Author: Addressed in https://github.com/blindnet-io/blindnet.dev/pull/40

This includes the classical consultation (access), deletion and modification of their data, of course, but also more advanced operations, including but not limited to:

- requesting the portability of their data to a compatible system
- advanced management of their consents (by objecting to and restricting proceeding of their data, or revoking previously given consents)
- and retrieving more generic privacy-related information regarding your organization (via the "transparency" request type).

:::note

> In accordance with GDPR and other legislations, these are only requests. **You** determine how those are handled and answered, manually or automatically.

Reviewer: ```suggestion
These are just requests. **You** determine how those are handled and answered, manually or automatically.
```

Author: Addressed in https://github.com/blindnet-io/blindnet.dev/pull/40

To add the Data Consumer Interface Web Component to your project, simply follow the same steps as for the Privacy Request Capture Interface one, but with the `@blindnet/dci` npm package, and `<bldn-data-consum>` HTML custom element.

```html
<!DOCTYPE html>
<html lang="en-GB">
  <head>
    <meta charset="utf-8" />
    <title>Privacy Components for the Web - Minimal Static Demo</title>
  </head>
  <body>
    <h1>Privacy Components for the Web - Minimal Static Demo</h1>

    <h2>Assert Your Privacy Rights</h2>

    <bldn-priv-request></bldn-priv-request>

    <h2>Manage Your Users' Requests</h2>

    <bldn-data-consum></bldn-data-consum>

    <script src="https://unpkg.com/@blindnet/prci/dist/index.core.min.js?module" type="module" ></script>
    <script src="https://unpkg.com/@blindnet/dci/dist/index.core.min.js?module" type="module" ></script>
  </body>
</html>
```

<!-- TODO: add link to API doc and more details -->

## 5. Answer Your Users' Requests

> :construction_worker: redaction in progress

Open your web app / page, and follow the instructions given in the Data Consumer Interface Web Component.

## 6. Set-up Storage

> > :construction_worker: redaction in progress

Reviewer: ```suggestion
> :construction_worker: writing in progress
```
redaction in EN means "the censoring or obscuring of part of a text for legal or security purposes."

Please make the same change elsewhere

Author: Addressed in #40
